lib/gemini.c

```diff
@@ -0,0 +1,267 @@
+{
+  return Curl_ssl_connect_nonblocking(conn, FIRSTSOCKET, done);
+}
+
+static CURLcode gemini_do_it(struct connectdata *conn, bool *done)
+{
+  struct GEMINI *gemini;
+  struct Curl_easy *data;
+  char *request;
+
+  data = conn->data;
+  request = aprintf("%s\r\n", data->change.url);
+
+  if(!request)
+    return CURLE_OUT_OF_MEMORY;
+
+  gemini = data->req.p.gemini;
+  gemini->request.memory = request;
+  gemini->request.amount_total = strlen(request);
+  gemini->request.amount_sent = 0;
+
+  /* Real work happens in gemini_doing, so we can use non-blocking
+   * functions and avoid busy loops.
+   */
+
+  return CURLE_OK;
+}
+
+static CURLcode gemini_doing_finish(struct connectdata *, bool *);
+static CURLcode gemini_doing(struct connectdata *conn, bool *done)
+{
+  CURLcode err;
+  curl_socket_t sockfd;
+  struct GEMINI *gemini;
+  size_t more;
+  size_t sent;
+  size_t amount;
+
+  gemini = conn->data->req.p.gemini;
+  sockfd = conn->sock[FIRSTSOCKET];
+
+  /* stage1: send request */
+  sent = gemini->request.amount_sent;
+  more = gemini->request.amount_total - sent;
+  if(more) {
+    char *from;
+
+    from = gemini->request.memory + sent;
+    err = Curl_write(conn, sockfd, from, more, &amount);
+    if(err)
+      return err;
+
+    gemini->request.amount_sent += amount;
+    more -= amount;
+
+    if(more)
+      return CURLE_OK;
+  }
+
+  /* stage2: read block big enough to contain header */
+  if(!gemini->block.done) {
+    char *into;
+
+    into = gemini->block.memory + gemini->block.amount;
+    more = GEMINI_RESPONSE_BUFSIZE - gemini->block.amount;
+
+    SOCKET_READABLE(sockfd, -1);
+    err = Curl_read(conn, sockfd, into, more, &amount);
+
+    if(err)
+      return err;
+
+    gemini->block.amount += amount;
+    more -= amount;
+
+    /* !more means that we succesfully read GEMINI_RESPONSE_BUFSIZE bytes.
+     * !amount means that there is no more data. It is quite possible
+     * for whole response, header + body combined to be less than
+     * GEMINI_RESPONSE_BUFSIZE bytes big.
+     */
+
+    if(!amount || !more)
+      gemini->block.done = TRUE;
+
+    /* Optimization: We check for LF, and skip reading more when it is
+     * found. Curl main engine adds noticable delays between
+     * invokactions of "doing" function, so it is desirable to get
+     * things done in as little calls to "doing" function, as possible,
+     * but without busy looping on socket that is not yet ready.
+     *
+     * For many servers first read returns exacly header, because it is
+     * natural thing to do on server side, although we can't rely on it.
+     * But this is reason why it does not worth to optimize search by
+     * keeping track of old {amount} value and searching only in bytes
+     * just read.
+     */
+    gemini->block.lf = memchr(gemini->block.memory, '\n',
+                              gemini->block.amount);
+    if(!gemini->block.lf && gemini->block.done)
+      return CURLE_WEIRD_SERVER_REPLY;
```

Comment:
maybe add a failf() call and explain the reason for the failure?

Reply:
Added explanation for every instance of CURLE_WEIRD_SERVER_REPLY.